### ISSUE TITLE
Question: Can you share your CI/CD setup?

### ISSUE THREAD

**Issue by FixRM** [WaelHamze/xrm-ci-framework#232]
Hello guys,

Can you share how do you do CI/CD on your D365 projects? I understand how the tools are working, but like to know a successful implementation story. Main areas of interest are:

- Dev environment setup: one env. per developer or shared? Or may be there are environments for every feature branch?
- Git branching model + workflow. Merging customization may be a nightmare so I guess it is always better to use one of the trunk-based models?
- Build/Test/ect. environment setup  + workflow. Are changes tested on dev env before pushing PR's? Do developer publish his changes to shared environment?
- CI/CD setup. What branch is used for? Do you test solution import for PR's or on merge? How do you update version, what is release flow?

So I'm interested in everything around the tools :) Also it will be cool to know about challenges and how do you handle them. May be somebody has "never do it this way" story? Any help is appreciated.

**Comment by mgr1054** [comment]
Hey, you might want to check out this Whitepaper published by Micrsoft. It describes a lot of patterns and approaches for D365 Solution Lifecycles. 

[https://www.microsoft.com/en-us/download/confirmation.aspx?id=57777](url)

**Comment by FixRM** [comment]
Thank you @mgr1054, I thought it was useless document I already saw, but it looks like it is updated version. Thank you!

**Comment by mgr1054** [comment]
The other day I had an idea for a deplyoment trigger. What do you guys think about implementing a Custom Button in Dynamics 365, where you would take the Id of the selected Solution and call the Azure DevOps REST API in order to start the build Pipeline? It seems pretty convinient to me.

**Comment by FixRM** [comment]
@mgr1054, I don't think it is a good idea. The whole idea behind CI/CD is to store solution in repo and be able to build any version at any time. But you are talking about automating of moving solution between environments. There is XrmToolbox plugin for that already, so you definitely don't need something in CRM or TFS for this

**Comment by mgr1054** [comment]
My goal is to minimize the effort for a developer (developing online by clicking in the ui of crm dynamics365) of deploying a new patch from dev to stage. It's hard to go the common way of ci, if there's no interaction between the developer and the repo in this use case.

**Comment by FixRM** [comment]
It's clear but Azure DevOps is designed to be code-driven. As already mentioned you it is easier to XrmToolbox or PS scripts for this task than involve Azure DevOps here.

**Comment by malaker** [comment]
In our case we have one shared development environment where we doing customizations. Once customization to the unmanaged solution is done then developer is extracting changes to code via packager. You can extract and unpack changes to code using the tool. Since we work in shared environment developer has to pick his changes and then commit to feature branch - its needed because not necessarily other developers finish the work.

From code repository we build manage solution by firstly copy empty environment to "stage" environment. For this part we have custom code written. Microsoft has something called Microsoft Xrm OnlineManagement Api that can be use to automate that part. Then to stage environment we import unmanaged solution. After that we can export managed solution which can be install on target environment.

**Comment by don-matese** [comment]
I have a shared environment and hooked in listener on the Solution Version number change event.  That kicks off an export, unpack (into source control), build and deploy to test
